README.md
# Clickstream Faker Provider for Python

- [Clickstream Faker Provider for Python](#clickstream-faker-provider-for-python)
  * [Purpose](#purpose)
  * [Installation](#installation)
  * [Usage](#usage)
  * [Features](#features)
  * [Contributing](#contributing)

## Purpose

This is a custom `Faker` provider for Python that generates clickstream session data. Data generated from this provider
represent user clickstream sessions on an online e-commerce site that sells mobile phones.

## Installation

The Clickstream Faker Provider for Python is available to install from PyPi using `pip`.

```bash
pip install faker_clickstream
```

## Usage

Sample code of Clickstream Provider usage.

```python
from faker import Faker
from faker_clickstream import ClickstreamProvider

fake = Faker()
fake.add_provider(ClickstreamProvider)
fake.session_clickstream()

# or...
fake.session_clickstream(rand_session_max_size=50)  # random number of events from 1 to 50
```

The `session_clickstream()` method returns an array of JSON objects that represents a unique web session. By default, is
configured to return a random number of session events from the range of 1 through 25.

An example response object is the below:

```json
[
  {
    "ip": "85.59.39.221",
    "user_id": 777198,
    "user_agent": "Mozilla/5.0 (iPad; CPU OS 7_1_2 like Mac OS X) AppleWebKit/537.51.2 (KHTML, like Gecko) CriOS/45.0.2454.68 Mobile/11D257 Safari/9537.53",
    "session_id": "d5cba274f7724780d1ed2b60650101892260748df21b0e2fb8b2b2fd88cedf23",
    "event_time": "28/03/2022 23:09:48.360212",
    "event_name": "AddToCart",
    "channel": "Social media",
    "metadata": {
      "product_id": 162012,
      "quantity": 1
    }
  },
  {
    "ip": "85.59.39.221",
    "user_id": 777198,
    "user_agent": "Mozilla/5.0 (iPad; CPU OS 7_1_2 like Mac OS X) AppleWebKit/537.51.2 (KHTML, like Gecko) CriOS/45.0.2454.68 Mobile/11D257 Safari/9537.53",
    "session_id": "d5cba274f7724780d1ed2b60650101892260748df21b0e2fb8b2b2fd88cedf23",
    "event_time": "28/03/2022 23:14:13.360227",
    "event_name": "AddToCart",
    "channel": "Social media",
    "metadata": {
      "product_id": 464710,
      "quantity": 2
    }
  },
  {
    "ip": "85.59.39.221",
    "user_id": 777198,
    "user_agent": "Mozilla/5.0 (iPad; CPU OS 7_1_2 like Mac OS X) AppleWebKit/537.51.2 (KHTML, like Gecko) CriOS/45.0.2454.68 Mobile/11D257 Safari/9537.53",
    "session_id": "d5cba274f7724780d1ed2b60650101892260748df21b0e2fb8b2b2fd88cedf23",
    "event_time": "28/03/2022 23:17:49.360241",
    "event_name": "Checkout",
    "channel": "Social media",
    "metadata": {}
  },
  {
    "ip": "85.59.39.221",
    "user_id": 777198,
    "user_agent": "Mozilla/5.0 (iPad; CPU OS 7_1_2 like Mac OS X) AppleWebKit/537.51.2 (KHTML, like Gecko) CriOS/45.0.2454.68 Mobile/11D257 Safari/9537.53",
    "session_id": "d5cba274f7724780d1ed2b60650101892260748df21b0e2fb8b2b2fd88cedf23",
    "event_time": "28/03/2022 23:22:15.360252",
    "event_name": "Search",
    "channel": "Social media",
    "metadata": {
      "query": "Samsung"
    }
  }
]
```

## Features

Each user session includes some static data that are randomly calculated only once, such as the `ip`, `user_agent`, etc.

The events that are available on this clickstream are the following ones, and an event dependency has been configured to
provide a more real representation of a web session flow:

- `Search`
- `AddToCart`
- `DeleteFromCart` (depends on `AddToCart` event)
- `IncreaseQuantity`
- `DecreaseQuantity` (depends on `IncreaseQuantity` event)
- `AddPromoCode` (depends on `Checkout` event)
- `Checkout` (depends on `AddToCart` event)
- `Login`
- `Logout` (depends on `Login` event)
- `CheckoutAsGuest`
- `CompleteOrder` (depends on **any** of `Checkout` or `CheckoutAsGuest` events)
- `CheckOrderStatus` (depends on `Login` event)

Moreover, the `Search`, `AddToCart`, `IncreaseQuantity` `DeleteFromCart` & `CheckOrderStatus` include a metadata field
with additional information about the products and the user queries.

## Contributing

Feel free to support this project and contribute! Support this project by giving a star!

CHANGELOG.md
# Changelog

## 0.1.1 (2022-04-02)

Features:

  - Include more IP addresses in the list.
  - Add default `Unknown` label on products with empty `os` info.
  
## 0.1.0 (2022-03-29)

Initial release.


tests/test_clickstream.py
import pytest


@pytest.fixture
def fake():
    from faker import Faker
    from faker_clickstream import ClickstreamProvider

    fake = Faker()
    fake.add_provider(ClickstreamProvider)
    fake.session_clickstream()
    return fake


def test_response_length(fake):
    res = fake.session_clickstream()
    assert len(res) > 0


def test_schema(fake):
    res = fake.session_clickstream()
    assert 'ip' in res[0]


faker_clickstream/ip.py
ip_list = [
    "58.217.40.197",
    "207.117.144.54",
    "65.169.228.128",
    "6.84.3.236",
    "214.56.36.93",
    "9.43.145.245",
    "231.199.26.76",
    "86.189.47.78",
    "3.122.58.75",
    "251.133.143.149",
    "6.142.181.83",
    "2.55.191.86",
    "90.189.58.233",
    "57.3.139.111",
    "41.130.99.194",
    "1.59.123.14",
    "9.119.131.109",
    "54.240.231.9",
    "117.19.131.144",
    "221.217.39.211",
    "7.43.125.76",
    "224.86.233.79",
    "32.151.38.98",
    "4.134.40.92",
    "4.70.188.58",
    "24.7.152.228",
    "58.122.179.1",
    "5.123.9.44",
    "3.175.206.5",
    "8.142.119.47",
    "40.248.58.203",
    "7.84.254.187",
    "215.215.239.71",
    "124.40.66.196",
    "215.87.143.102",
    "143.39.97.13",
    "202.56.79.198",
    "143.60.125.142",
    "73.233.153.126",
    "121.144.28.245",
    "53.232.193.122",
    "104.222.102.209",
    "216.55.215.13",
    "84.106.145.239",
    "200.131.52.20",
    "132.252.158.0",
    "220.11.129.27",
    "8.153.0.186",
    "119.206.117.107",
    "222.93.42.133",
    "105.104.224.98",
    "252.156.181.78",
    "7.139.235.187",
    "111.140.127.91",
    "1.186.17.84",
    "85.59.39.221",
    "231.152.252.43",
    "125.214.216.123",
    "69.31.65.238",
    "126.12.105.55",
    "211.39.4.118",
    "73.102.88.79",
    "210.229.38.137",
    "5.3.230.214",
    "208.92.91.242",
    "4.105.177.199",
    "38.10.48.92",
    "133.33.44.13",
    "202.189.24.255",
    "5.101.244.234",
    "2.52.110.194",
    "1.130.73.146",
    "84.237.232.120",
    "25.163.83.194",
    "3.190.95.171",
    "44.144.252.137",
    "7.246.182.101",
    "90.28.208.71",
    "249.135.140.63",
    "217.174.57.192",
    "4.209.187.40",
    "147.177.235.135",
    "100.125.39.11",
    "88.217.209.75",
    "131.178.173.24",
    "7.106.250.95",
    "57.116.146.168",
    "66.249.236.49",
    "215.238.68.242",
    "5.195.182.250",
    "54.16.251.155",
    "209.129.140.18",
    "56.151.31.174",
    "220.119.182.223",
    "9.253.207.162",
    "21.239.61.163",
    "8.161.122.98",
    "215.235.153.76",
    "233.254.145.198",
    "125.102.121.103",
    "160.91.40.89",
    "154.113.159.58",
    "155.244.103.67",
    "192.189.53.130",
    "4.5.140.190",
    "245.83.76.99",
    "14.196.103.255",
    "104.29.7.39",
    "115.217.86.42",
    "2.163.94.50",
    "205.246.117.121",
    "62.223.113.227",
    "59.144.235.145",
    "104.168.216.219",
    "99.237.164.97",
    "159.212.115.41",
    "64.126.143.120",
    "108.10.11.103",
    "41.144.34.11",
    "174.152.141.214",
    "4.184.215.7",
    "183.173.216.59",
    "93.80.231.220",
    "66.64.147.81",
    "71.83.196.176",
    "255.234.85.245",
    "45.40.49.43",
    "201.6.90.55",
    "32.65.183.136",
    "19.251.19.245",
    "118.64.82.139",
    "161.246.115.136",
    "29.37.87.149",
    "60.184.209.0",
    "146.83.175.103",
    "42.214.67.8",
    "236.127.226.169",
    "29.52.53.183",
    "5.59.163.232",
    "183.212.131.95",
    "123.73.135.147",
    "212.167.210.48",
    "247.53.174.177",
    "235.13.153.145",
    "214.186.83.149",
    "227.93.229.145",
    "131.1.8.207",
    "66.108.14.68",
    "66.221.183.7",
    "135.125.143.15",
    "255.245.114.116",
    "142.192.177.236",
    "225.105.189.32",
    "70.223.114.82",
    "114.155.19.170",
    "24.24.46.88",
    "63.181.186.170",
    "167.92.29.1",
    "124.190.125.200",
    "114.117.7.83",
    "251.221.185.234",
    "113.151.151.240",
    "209.20.248.182",
    "202.107.146.213",
    "48.22.16.86",
    "55.254.203.155",
    "229.170.207.196",
    "254.167.4.89",
    "187.127.145.64",
    "244.33.77.99",
    "131.214.19.153",
    "162.17.189.171",
    "230.91.71.143",
    "223.4.223.54",
    "210.40.105.214",
    "153.194.135.41",
    "101.174.74.92",
    "181.148.218.147",
    "182.58.180.204",
    "201.151.181.148",
    "232.176.113.46",
    "154.32.203.118",
    "250.182.199.16",
    "199.211.251.78",
    "238.235.169.167",
    "70.70.44.222",
    "220.50.186.233",
    "94.122.67.136",
    "125.30.23.254",
    "68.189.198.207",
    "203.92.37.229",
    "44.136.241.70",
    "196.82.148.71",
    "239.246.178.14",
    "230.87.29.155",
    "69.234.90.152",
    "19.226.113.96",
    "2.121.166.168",
    "162.8.235.103",
    "25.197.229.115",
    "227.6.30.177",
    "215.64.214.91",
    "104.49.106.184",
    "212.238.246.215",
    "131.32.124.127",
    "34.205.30.234",
    "231.63.74.141",
    "97.230.49.72",
    "99.94.173.203",
    "190.83.123.54",
    "86.165.143.183",
    "62.243.81.172",
    "129.217.167.211",
    "54.104.115.18",
    "14.111.83.115",
    "71.251.136.255",
    "233.229.115.64",
    "221.86.24.230",
    "112.182.248.136",
    "175.158.244.4",
    "129.131.58.189",
    "46.174.64.1",
    "18.54.45.122",
    "232.208.99.168",
    "28.243.137.204",
    "126.93.215.20",
    "77.187.12.94",
    "172.36.45.63",
    "124.186.185.58",
    "79.40.249.63",
    "93.126.156.221",
    "95.25.200.154",
    "252.54.85.252",
    "44.188.81.31",
    "131.5.208.167",
    "232.58.220.143",
    "180.162.247.191",
    "209.193.139.26",
    "243.147.247.45",
    "50.193.44.33",
    "245.148.12.105",
    "98.26.195.74",
    "237.51.148.245",
    "130.121.182.26",
    "243.167.79.57",
    "72.234.168.114",
    "65.240.35.245",
    "186.15.253.102",
    "119.251.231.24",
    "223.218.80.36",
    "62.110.7.44",
    "145.158.152.224",
    "181.219.68.52",
    "244.114.99.160",
    "222.33.181.63",
    "7.126.10.90",
    "58.43.27.198",
    "225.227.163.185",
    "5.104.237.189",
    "77.143.165.82",
    "47.143.90.206",
    "199.45.47.74",
    "168.0.10.129",
    "82.46.164.122",
    "225.177.4.55",
    "153.180.86.84",
    "216.113.77.64",
    "6.211.110.142",
    "151.0.240.147",
    "121.150.149.208",
    "234.200.186.8",
    "44.62.9.126",
    "7.226.23.130",
    "215.24.42.14",
    "171.116.250.97",
    "140.51.147.224",
    "91.127.130.248",
    "28.53.229.150",
    "14.44.108.146",
    "0.94.248.36",
    "144.226.191.61",
    "238.115.20.253",
    "131.187.237.250",
    "80.110.96.123",
    "109.65.33.125",
    "225.176.243.185",
    "251.145.66.238",
    "28.86.112.12",
    "184.140.84.190",
    "93.43.25.243",
    "215.48.4.101",
    "57.134.103.103",
    "78.1.137.65",
    "176.182.123.142",
    "49.224.159.158",
    "240.165.166.227",
    "30.14.197.190",
    "89.66.145.111",
    "90.58.164.58",
    "215.248.9.169"
]

response_codes = [
    "200",
    "302",
    "404",
    "405",
    "406",
    "407"
]


faker_clickstream/event_constants.py
events = [
    "Search",
    "AddToCart",
    "DeleteFromCart",
    "IncreaseQuantity",
    "DecreaseQuantity",
    "AddPromoCode",
    "Checkout",
    "Login",
    "Logout",
    "CheckoutAsGuest"
    "CompleteOrder",
    "CheckOrderStatus"
]

weighted_events = [
    {
        "name": "Search",
        "popularity": 85,
        "dependsOn": [],
        "dependencyFilter": "any"
    },
    {
        "name": "AddToCart",
        "popularity": 70,
        "dependsOn": [],
        "dependencyFilter": "any"
    },
    {
        "name": "DeleteFromCart",
        "popularity": 55,
        "dependsOn": ["AddToCart"],
        "dependencyFilter": "any"
    },
    {
        "name": "IncreaseQuantity",
        "popularity": 40,
        "dependsOn": [],
        "dependencyFilter": "any"
    },
    {
        "name": "DecreaseQuantity",
        "popularity": 40,
        "dependsOn": ["IncreaseQuantity"],
        "dependencyFilter": "any"
    },
    {
        "name": "AddPromoCode",
        "popularity": 25,
        "dependsOn": ["Checkout"],
        "dependencyFilter": "any"
    },
    {
        "name": "Checkout",
        "popularity": 75,
        "dependsOn": ["AddToCart"],
        "dependencyFilter": "any"
    },
    {
        "name": "Login",
        "popularity": 70,
        "dependsOn": [],
        "dependencyFilter": "any"
    },
    {
        "name": "Logout",
        "popularity": 30,
        "dependsOn": ["Login"],
        "dependencyFilter": "any"
    },
    {
        "name": "CheckoutAsGuest",
        "popularity": 40,
        "dependsOn": [],
        "dependencyFilter": "any"
    },
    {
        "name": "CompleteOrder",
        "popularity": 65,
        "dependsOn": ["Checkout", "CheckoutAsGuest"],
        "dependencyFilter": "any"
    },
    {
        "name": "CheckOrderStatus",
        "popularity": 70,
        "dependsOn": ["Login"],
        "dependencyFilter": "any"
    }
]

channel = [
    "Organic search",
    "Direct",
    "Social media",
    "Referral",
    "Other"
]


faker_clickstream/clickstream.py
import hashlib
import random
import string
from datetime import datetime, timedelta
from random import choice, randint

from faker.providers import BaseProvider

from faker_clickstream.event_constants import weighted_events, events, channel
from faker_clickstream.ip import ip_list
from faker_clickstream.mobile_phones import mobile_phones
from faker_clickstream.user_agents import user_agents


class ClickstreamProvider(BaseProvider):
    """
        A Provider for clickstream related test data.

        >>> from faker import Faker
        >>> from faker_clickstream import ClickstreamProvider
        >>> fake = Faker()
        >>> fake.add_provider(ClickstreamProvider)
        >>> fake.session_clickstream()
    """

    def user_agent(self):
        """
        Generate random user agent.

        :return: User agent string
        """
        return choice(user_agents)

    def event(self):
        """
        Generate random event type name for e-commerce site.

        :return: Event type string
        """
        return choice(events)

    def weighted_event(self):
        """
        Generate a random event object according to popularity weight. Higher popularity increases the
        chances of occurrence.

        :return: Event object (JSON)
        """
        return random.choices(weighted_events, weights=[e['popularity'] for e in weighted_events], k=1)[0]

    def session_clickstream(self, rand_session_max_size: int = 25):
        """
        Generate session clickstream events.

        :param rand_session_max_size: Max number of possible events in session. Defaults to 25.
        :return: List of session events
        """

        # Initialize static session values
        session_events = []
        user_id = _get_user_id()
        user_agent = self.user_agent()
        session_id = _get_session_id()
        ip = _get_ip()
        channel_type = _get_channel()
        random_session_size = randint(1, rand_session_max_size)
        incremental_delta_delay = randint(1, 60)

        # Keep track of unique values in a session
        unique_session_events = set()
        product_codes = set()

        for s in range(random_session_size):
            # Mock time delay between events
            incremental_delta_delay = incremental_delta_delay + (s * randint(1, 60))
            event_time = _format_time(_get_event_time(delta=incremental_delta_delay))

            # Fetch weighted event
            event = self.weighted_event()

            if (event['name'] == 'Login' and event['name'] in unique_session_events) \
                    or (event['name'] == 'CheckoutAsGuest' and user_id != 0):
                # If user ID is not 0, discard CheckoutAsGuest event
                # or Login exists in session, discard Login event
                # Add a mock Search event
                event['name'] = 'Search'

            if event['name'] == 'Login' and user_id == 0:
                # If user id is -1 and Login event, regenerate user ID.
                user_id = _get_user_id(start=1)

            if (event['name'] == 'Login' and user_id != 0) or (event['name'] == 'Logout' and user_id == 0):
                # Add a mock Search event
                event['name'] = 'Search'

            # Keep track of unique events in session
            unique_session_events.add(event['name'])

            # Handle event dependencies
            if len(event['dependsOn']):
                list_check = [d in unique_session_events for d in event['dependsOn']]
                if event['dependencyFilter'] == 'all':
                    f = all(list_check)
                else:
                    f = any(list_check)
                if not f:
                    # Add a mock Search event
                    event['name'] = 'Search'

            # If CompleteOrder, remove some events from the unique list to reoccur.
            if event['name'] == 'CompleteOrder':
                if 'Checkout' in unique_session_events:
                    unique_session_events.remove('Checkout')
                if 'CheckoutAsGuest' in unique_session_events:
                    unique_session_events.remove('CheckoutAsGuest')
                if 'DecreaseQuantity' in unique_session_events:
                    unique_session_events.remove('DecreaseQuantity')

            # Fill metadata object conditionally
            metadata = {}
            if event['name'] == 'Search':
                sample_product = _get_weighted_mobile_phone()
                metadata['query'] = choice(
                    (sample_product['model_name'], sample_product['brand_name'], sample_product['os'])
                )

            if event['name'] in ('AddToCart', 'IncreaseQuantity'):
                metadata['product_id'] = _get_product_code()
                metadata['quantity'] = _get_quantity()
                product_codes.add(metadata['product_id'])

            if event['name'] == 'DeleteFromCart':
                if len(product_codes):
                    random_delete = choice(list(product_codes))
                    product_codes.remove(random_delete)
                    metadata['product_id'] = random_delete

            if event['name'] == 'CheckOrderStatus':
                metadata['order_id'] = _get_order_id()

            # Construct final event object
            r = {
                "ip": ip,
                "user_id": user_id,
                "user_agent": user_agent,
                "session_id": session_id,
                "event_time": event_time,
                "event_name": event['name'],
                "channel": channel_type,
                "metadata": metadata
            }
            session_events.append(r)
        return session_events


def _get_session_id():
    """
    Generate session ID

    :return: Session ID string
    """
    return hashlib.sha256(
        ('%s%s%s' % (
            datetime.now().strftime("%d/%m/%Y %H:%M:%S.%f"),
            (''.join(random.choice(string.ascii_lowercase)) for _ in range(10)),
            'faker_clickstream'
        )).encode('utf-8')
    ).hexdigest()


def _get_product_code():
    """
    Generate random product code from range 1 to 999999.

    :return: Random integer number
    """
    return randint(1, 999999)


def _get_order_id():
    """
    Generate random order id from range 1 to 999999.

    :return: Random integer number
    """
    return randint(1, 999999)


def _get_user_id(start: int = 0, end: int = 999999):
    """
    Generate random user id from range 0 to 999999. Zero value may identify null user.

    :param start: Index start (Default: 0)
    :param end: Index end (Default: 999999)
    :return:
    """
    return randint(start, end)


def _get_event_time(delta):
    """
    Generate current event time, added by some delta value.

    :param delta: Delta time value in seconds
    :return: Event time
    """
    return datetime.now() + timedelta(seconds=delta)


def _format_time(t):
    """
    Format time to string.

    :param t: Time object
    :return: Time string in format like 28/03/2022 23:22:15.360252
    """
    return t.strftime("%d/%m/%Y %H:%M:%S.%f")


def _get_quantity():
    """
    Get random product order quantity from 1 to 5. Values are given a weight, decreasing as the quantity number
    increases.

    :return: Product quantity number
    """
    return random.choices([1, 2, 3, 4, 5], weights=[50, 20, 20, 5, 5], k=1)[0]


def _get_weighted_mobile_phone():
    """
    Get mobile phone object according to popularity

    :return: Mobile phone object
    """
    return random.choices(mobile_phones, weights=[e['popularity'] for e in mobile_phones], k=1)[0]


def _get_ip():
    """
    Get random IP address from list.

    :return: IP address string
    """
    return choice(ip_list)


def _get_channel():
    """
    Get user origin channel (e.g. "Organic search", "Direct", "Social media", "Referral", "Other")

    :return: Origin channel string
    """
    return choice(channel)
